### PR TITLE
QtPy, Python 3, upstream dependencies

### PR DESCRIPTION
**Pull request by bgermann**
This makes graide compatible with QtPy and Python 3.
QtPy forwards the Qt5 API calls to one of four backends, including PySide and the Qt5-based PySide2.
Python 2 is currently phased out on many distributions because its end of life is scheduled on 1.1.2020.

The vendored, old freetype-py dependency is removed and replaced by a dependency on the upstream module that is now available with binaries on PyPI.
Additionally, the graide.graphite module is replaced with the upstream graphite2 module. This feature depends on https://github.com/silnrsi/graphite being published on PyPI.

Furthermore, a feature to compile grcompiler on creating a wheel is introduced.

### CODE REVIEW

**Comment by mhosken** [comment]
May I ask what is to be gained by switching to qtpy over pyside?

**Comment by bgermann** [comment]
On the [PySide source repository](https://github.com/pyside/PySide) there is a big warning "ATTENTION: This project is deprecated, please refer to PySide2." The main reason is that PySide  is a Qt4 binding and Qt4 is end of life and currently phased out by many Linux distributions.
So why not use PySide2? I chose QtPy because you can use PySide or PySide2 as a backend. So people that absolutely want to stick to Qt4 can use PySide. And it was easier to port as I could just change the QT_API environment variable to use either PySide or PySide2.
However QtPy has exactly the same API as PySide2 so the only difference are the import statements and those can be changed quite easily.

**Comment by bgermann** [comment]
I think this is ready to go now. It would make sense to stop exporting installation files with PyInstaller and to publish graide on PyPI instead. Then a user would be able to install graide via `pip install graphite-graide` and `pip install PySide(2)` on each supported platform. Renaming the package to graphite2-graide or just graide would make sense because there is an unrelated graphite package on PyPI already.

File: lib/graide/gdx.py [hunk 1]
```diff
@@ -83,8 +83,8 @@ def __init__(self, e, relbase) :
...
         self.pretty = e.get('prettyPrint')
         slots = map(lambda x: int(x.get('slotIndex')), e.findall('rhsSlot'))
-        if len(slots) :
-            d = slots[0]
+        if any(slots) :
+            d = next(slots)
             self.slots = map(lambda x : x - d, slots)
         else :
-            self.slots = []
+            self.slots = {}
```
Comment (mhosken) [inline lib/graide/gdx.py]: why change from empty list to empty dict here?

File: lib/graide/run.py [hunk 2]
```diff
@@ -37,12 +37,12 @@ def addSlots(self, runinfo, flipDir = False) :
...
     # Reverse the slots, but keeping diacritics (directionality = 16) after their bases.
     def reverseDirection(self) :
-        #print "reverseDirection"
+        #print("reverseDirection")
         #self.printDebug()
         runTemp = Run(self.font, self.rtl)
         #print "runTemp length=",len(runTemp)
         iLastOfSeq = len(self)
-        for i in xrange(len(self)-1, 0, -1):
+        for i in list(range(len(self)-1, 0, -1)):
```
Comment (mhosken) [inline lib/graide/run.py]: No need for a list here?

File: graide.spec [hunk 3]
```diff
@@ -34,17 +32,15 @@ if sys.platform == 'win32' :
...
              hookspath=None)
 pyz = PYZ(a.pure)
 bins = a.binaries
+for d in glob.glob('lib/graide/grcompiler/*.*') :
+    bins += [(d[22:], d, 'BINARY')]
 if sys.platform == 'win32' :
-    for d in ('zlib1', 'freetype6', 'graphite2', 'msvcr100') :
-        bins += [(d + '.dll', 'build/scripts-2.7/' + d + '.dll', 'BINARY')]
-    for d in glob.glob('grcompiler_win/*.*') :
-        bins += [(d[15:], d, 'BINARY')]
```
Comment (mhosken) [inline graide.spec:47]: If not here, where do these dependencies get pulled in? (Likewise for Darwin)

File: lib/graide/glyph.py [hunk 4]
```diff
@@ -50,9 +50,8 @@ class GlyphItem(object) :
     def __init__(self, face, gid, height = 40) :
         face.set_char_size(height = int(height * 64))
         (self.pixmap, self.left, self.top) = ftGlyph(face, gid)
-        n = ctypes.create_string_buffer(64)
-        freetype.FT_Get_Glyph_Name(face._FT_Face, gid, n, ctypes.sizeof(n))
-        self.name = re.sub(ur'[^A-Za-z0-9._]', '', n.value) # Postscript name
+        name = face.get_glyph_name(gid).decode('ascii')
+        self.name = re.sub('[^A-Za-z0-9._]', '', name) # Postscript name
```
Comment (mhosken) [inline lib/graide/glyph.py:54]: Where is the implementation of get_glyph_name() and load_glyph()

**Comment by mhosken** [comment]
Initial read through. Things look good. A couple of questions/issues:

. Does this still run fine on python2.7. I noticed the loss of a couple of unicode() calls and switching isinstance(x, basestring) to isinstance(x, str) may cause problems.
. Losing the bundling of dependencies for Windows users could be a bit of a pain. I suppose pip solves all that now?

File: graide.spec (hunk 3 above)
Comment (bgermann) [reply graide.spec:47]: In setup.py there is install_requires which makes setuptools pull them in if they are released on PyPI which is the case for all dependencies except graphite2. If you merge silnrsi/graphite#37 I volunteer to publish the Python package. The setuptools will install binary wheels if there are any for the platform in question. All dependencies' Python packages that depend on dynamic libraries have binary wheels available for Windows, macOS and Linux. Those are [freetype-py](https://pypi.org/project/freetype-py/#files) and [PySide2](https://pypi.org/project/PySide2/#files); [PySide](https://pypi.org/project/PySide/#files) only has Windows binaries.

I added grcompiler.py, enabling setuptools to download and compile grcompiler for each supported platform with GRCOMPILER_BUNDLE_VERSION environment variable set. Only the person publishing the graide package on PyPI would have to use that.

File: lib/graide/glyph.py (hunk 4 above)
Comment (bgermann) [reply lib/graide/glyph.py:54]: https://github.com/rougier/freetype-py/blob/master/freetype/__init__.py

File: lib/graide/gdx.py (hunk 1 above)
Comment (bgermann) [reply lib/graide/gdx.py]: A dict is more aligned with what map() returns.

File: lib/graide/run.py (hunk 2 above)
Comment (bgermann) [reply lib/graide/run.py]: Right.

**Comment by bgermann** [comment]
I made sure this still runs on Python 2.7. That is what future's builtins.{chr,str} are for.

Pip solves all dependencies, even on Windows. In fact, all popular Python projects are moving from bundling the Python interpreter and dependencies incl. binaries in a setup.exe to using wheels. That has the advantage of the user being in control of the Python versions that they have installed with their potential security issues.

**Comment by devosb** [comment]
@bgermann I admire the carefulness to not break Python 2.7 support. @mhosken Why do we need Python 2.7 compatibility?

**Comment by mhosken** [comment]
OK I've started a more in depth review of this PR.

1. `GRCOMPILER_BUNDLE_VERSION=master python setup.py build` doesn't build grcompiler. It only builds for bdist_wheel (which is an undocumented setup.py according to my setup.py --help-commands).

2. Please don't make people have to install a toolchain for grcompiler just to use graide from source. If you want to build a fresh grcompiler during bdist, that's fine, but not just to let people run the thing. The same applies to the other .dlls. Could we make it that if you do setup.py without setting GRCOMPILER_BUNDLE_VERSION, it uses the bundled version otherwise it builds it?

3. ttfrename should not be bundled as a separate app. It's old test code and isn't a releasable app. It is also wrongly named since it clashes with ttfrename which ships with the fontutils package (yeh it's written in perl, but it still is a valid tool on people's systems). Best just ignore that it is there and I should probably remove it.

Please bare in mind that a primary use case is not pip based but Ubuntu package based. So any dependencies must already be available as debian packages and not just in pip. I'm OK with the bdist process for pip shedding supporting libraries in favour of cross pip dependencies, but I would still like to see the supporting libraries left in the repo, at least for now.

Having said all this. I do like the idea of people installing using pip rather than an installer .exe. I'm fine with pyinstaller going away.

I'm intrigued why you got drawn into working on graide. It's a bit of a corner product.

**Comment by bgermann** [comment]
The three changes that you suggested are in the PR now. By default, the binaries are not bundled (as this is what you want for most setup() calls. For bdist* one can set GRCOMPILER_BUNDLE to an arbitrary value to get the prebuilt version).

The missing Debian packages are python-freetype and python-graphite2.
I am building a [python-freetype Debian package](https://salsa.debian.org/python-team/modules/freetype-py) (not published yet) and extended PR https://github.com/silnrsi/graphite/pull/37 to include Debian python packages.

File: lib/graide/gdx.py (hunk 1 above)
Comment (tim-eves) [reply lib/graide/gdx.py]: I'm pretty sure map on python2 returns a list and on python3 an iterator, assuming we're talking about the standard function

**Comment by bgermann** [comment]
Thank you for merging this. Maybe you want to take the last two commits as well.

**Pull request closed without merging**